python/semantic_kernel/connectors/memory/azure_ai_search/azure_ai_search_memory_record_service.py

```diff
@@ -0,0 +1,432 @@
+
+        Args:
+            collection_name (str): The name of the collection to get the record from.
+            key (str): The unique database key of the record.
+            with_embedding (bool): Whether to include the embedding in the result. (default: {False})
+
+        Returns:
+            MemoryRecord: The record.
+        """
+        # Look up Search client class to see if exists or create
+        search_client = self._search_index_client.get_search_client(collection_name.lower())
+
+        try:
+            search_result = await search_client.get_document(
+                key=encode_id(key), selected_fields=get_field_selection(with_embedding)
+            )
+        except ResourceNotFoundError as exc:
+            await search_client.close()
+            raise MemoryConnectorResourceNotFound("Memory record not found") from exc
+
+        await search_client.close()
+
+        # Create Memory record from document
+        return self._item_type(**search_result)
+
+    @overload
+    async def get_batch(
+        self, collection_name: str, keys: list[str], with_embeddings: bool = False
+    ) -> list[MemoryRecord]:
+        """Gets a batch of records.
+
+        Args:
+            collection_name (str): The name of the collection to get the records from.
+            keys (List[str]): The unique database keys of the records.
+            with_embeddings (bool): Whether to include the embeddings in the results. (default: {False})
+
+        Returns:
+            List[MemoryRecord]: The records.
+        """
+        search_results = []
+
+        for key in keys:
+            search_result = await self.get(
+                collection_name=collection_name.lower(),
+                key=key,
+                with_embedding=with_embeddings,
+            )
+            search_results.append(self._item_type(**search_result))
+
+        return search_results
+
+    @overload
+    async def delete_batch(self, collection_name: str, keys: list[str]) -> None:
+        """Removes a batch of records.
+
+        Args:
+            collection_name (str): The name of the collection to remove the records from.
+            keys (List[str]): The unique database keys of the records to remove.
+
+        Returns:
+            None
+        """
+        for record_id in keys:
+            await self.remove(collection_name=collection_name.lower(), key=encode_id(record_id))
+
+    @overload
+    async def delete(self, collection_name: str, key: str) -> None:
+        """Removes a record.
+
+        Args:
+            collection_name (str): The name of the collection to remove the record from.
+            key (str): The unique database key of the record to remove.
+
+        Returns:
+            None
+        """
+        # Look up Search client class to see if exists or create
+        search_client = self._search_index_client.get_search_client(collection_name.lower())
+        docs_to_delete = {SEARCH_FIELD_ID: encode_id(key)}
```

Comment:
Opinionated behaviors like encoding an id will need to be made optional or injectable in the new stack. In order to make it possible for users to use our system with existing data stores, it means we cannot change the id's they pass in, because it might cause the id to not match what is in the db.  We'll still need to be able to optionally do this kind of thing though, if we want to have an opinionated stack that is compatible with today's system available for backwards compatibility.  The way I'm thinking about that is to just use the new stack but injecting mappers or encoders that add the behaviors we have in the old stack, and making that available via some special factory methods.